### PR TITLE
Add option to prefer an Email address to a Username

### PR DESCRIPTION
**Pull request by Spindel**
## Description

Implement an option to prefer using email address as the username to upstream services.

## Motivation and Context

With some providers the Username is an upstream Unique ID, like fex. in the
case of Google, where you get a stable string of characters that are guaranteed to be unique for Google..

When matching this with downstream databases, it's sometimes preferred to use
the email address as the  known identifier.

However, when _mixing_ this with sometimes other sources, like htaccess, which
doesn't have a concept of an email address, it can turn difficult, as you can no longer only trust the "X-Forwarded-Email" header but must manage fallback handling, etc.

This change makes the headers _prefer_ to use the Email address, if such exists,
for the Username identifier when passing data to downstream services. 

Defaults to Off.

## How Has This Been Tested?

Unittests only

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

**Comment by Spindel** [comment]
Now it's been live tested as well.

File: docs/configuration/configuration.md [hunk 1]
```diff
@@ -74,6 +74,7 @@ An example [oauth2_proxy.cfg]({{ site.gitweb }}/contrib/oauth2_proxy.cfg.example
 | `-pass-access-token` | bool | pass OAuth access_token to upstream via X-Forwarded-Access-Token header | false |
 | `-pass-authorization-header` | bool | pass OIDC IDToken to upstream via Authorization Bearer header | false |
 | `-pass-basic-auth` | bool | pass HTTP Basic Auth, X-Forwarded-User and X-Forwarded-Email information to upstream | true |
+| `-prefer-email-to-user` | bool | pass Prefer to use the Email address as the Username when passing information to upstream. Will only use Username if Email is unavailable, eg. htaccess authentication. | false |
```
Comment (JoelSpeed) [inline docs/configuration/configuration.md]:
```suggestion
| `-prefer-email-to-user` | bool | Prefer to use the Email address as the Username when passing information to upstream. Will only use Username if Email is unavailable, eg. htaccess authentication. | false |
```

File: options.go [hunk 2]
```diff
@@ -279,6 +281,10 @@ func (o *Options) Validate() error {
 		}
 	}
 
+	if o.PreferEmailToUser == true && o.PassBasicAuth == false {
+		msgs = append(msgs, "Prefer Email To User is only useful with PassBasicAuth")
```
Comment (JoelSpeed) [inline options.go]:
```suggestion
		msgs = append(msgs, "PreferEmailToUser should only be used with PassBasicAuth")
```

**Comment by Spindel** [comment]
Thanks for the review, I'll fix that up on Monday and re-push my branch

**Comment by Spindel** [comment]
Squashed the changes up, should be all clear.

**Comment by Spindel** [comment]
Since there's some conflicts now vs. upstream/master, how should I proceed with this?  Do you want me to rebase my branch, or merge master into this?

**Comment by JoelSpeed** [comment]
Rebase would be preferable if that's ok

**Comment by Spindel** [comment]
Rebased the changes.

**Comment by Spindel** [comment]
So, rebasing drops the reviews as stale? Annoying.

**Pull request merged**